stdlib/mlang/symbolize.mc

```diff
@@ -0,0 +1,787 @@
+                    t.ident
+
+    sem symbolizeDecl : SymEnv -> Decl -> (SymEnv, Decl)
+    sem symbolizeDecl env = 
+    | DeclInclude r ->
+        error "Symbolization expects all DeclInclude to have been removed!"
+    | DeclLet t ->  
+        match symbolizeTyAnnot env t.tyAnnot with (tyVarEnv, tyAnnot) in
+        match setSymbol env.currentEnv.varEnv t.ident with (varEnv, ident) in
+        let env = updateVarEnv env varEnv in 
+        let env = updateTyVarEnv env tyVarEnv in 
+        let decl = DeclLet {t with ident = ident,
+                            tyAnnot = tyAnnot,
+                            body = symbolizeExpr env t.body} in 
+        (env, decl)
+    | DeclType t -> 
+        match setSymbol env.currentEnv.tyConEnv t.ident with (tyConEnv, ident) in
+        let env = updateTyConEnv env tyConEnv in 
+        match mapAccumL setSymbol env.currentEnv.tyVarEnv t.params with (tyVarEnv, params) in
+        let decl = DeclType {t with ident = ident,
+                             params = params,
+                             tyIdent = symbolizeType (updateTyVarEnv env tyVarEnv) t.tyIdent} in 
+        let env = updateTyVarEnv env tyVarEnv in 
+        (env, decl)
+    | DeclRecLets t -> 
+        -- Generate fresh symbols for all identifiers and add to the environment
+        let setSymbolIdent = lam env. lam b.
+            match setSymbol env b.ident with (env, ident) in
+            (env, {b with ident = ident})
+        in
+
+        match mapAccumL setSymbolIdent env.currentEnv.varEnv t.bindings with (varEnv, bindings) in
+        let newEnv = updateVarEnv env varEnv in
+
+        -- Symbolize all bodies with the new environment
+        let bindings =
+        map (lam b. match symbolizeTyAnnot env b.tyAnnot with (tyVarEnv, tyAnnot) in
+                    {b with body = symbolizeExpr (updateTyVarEnv newEnv tyVarEnv) b.body,
+                            tyAnnot = tyAnnot})  bindings in
+
+        (newEnv, DeclRecLets {t with bindings = bindings})
+    | DeclConDef t -> 
+        match setSymbol env.currentEnv.conEnv t.ident with (conEnv, ident) in
+
+        let decl = DeclConDef {t with ident = ident,
+                                      tyIdent = symbolizeType env t.tyIdent} in 
+        let env = updateConEnv env conEnv in 
+        (env, decl)
+    | DeclUtest t -> 
+        -- This can be rewritten to use a shallow map on declarations. E.g.
+        -- smap (symbolizeExpr env) (DeclUtest t) 
+        let decl = DeclUtest {t with test = symbolizeExpr env t.test,
+                                     expected = symbolizeExpr env t.expected,
+                                     tusing = optionMap (symbolizeExpr env) t.tusing} in 
+        (env, decl)
+    | DeclExt t -> 
+        -- When ignoreExternals is set, the DeclExt should have been filtered
+        -- out of the program before attempting to symbolize the declarations.
+        if env.ignoreExternals then
```

Comment:
Is this consistent with handling for externals in regular `mexpr`?

Reply:
Ill fix this